File: src/main/java/mods/cybercat/gigeresque/common/block/NestResinBlock.java
package mods.cybercat.gigeresque.common.block;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.state.properties.IntegerProperty;
import net.minecraft.world.level.pathfinder.PathComputationType;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.EntityCollisionContext;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

import mods.cybercat.gigeresque.common.entity.AlienEntity;

public class NestResinBlock extends Block {

    public static final IntegerProperty LAYERS = BlockStateProperties.LAYERS;

    protected static final List<VoxelShape> ALIEN_LAYERS_TO_SHAPE = interpolateShapes(false);

    protected static final List<VoxelShape> LAYERS_TO_SHAPE = interpolateShapes(true);

    public NestResinBlock(Properties settings) {
        super(settings);

        registerDefaultState(getStateDefinition().any().setValue(BlockStateProperties.LAYERS, 1));
    }

    private static List<VoxelShape> interpolateShapes(boolean divide) {
        ArrayList<VoxelShape> list = new ArrayList<>();
        list.add(Shapes.empty());
        for (var i = 0; i < 8; i++)
            list.add(box(0.0, 0.0, 0.0, 16.0, divide ? (i * 2.0) / 2.0 : i * 2.0, 16.0));
        return list;
    }

    @Override
    public boolean isPathfindable(
        @NotNull BlockState state,
        @NotNull BlockGetter world,
        @NotNull BlockPos pos,
        @NotNull PathComputationType type
    ) {
        return type == PathComputationType.LAND;
    }

    @Override
    public @NotNull VoxelShape getShape(
        BlockState state,
        @NotNull BlockGetter world,
        @NotNull BlockPos pos,
        @NotNull CollisionContext context
    ) {
        return ALIEN_LAYERS_TO_SHAPE.get(state.getValue(LAYERS));
    }

    @Override
    public @NotNull VoxelShape getCollisionShape(
        @NotNull BlockState state,
        @NotNull BlockGetter world,
        @NotNull BlockPos pos,
        @NotNull CollisionContext context
    ) {
        return context instanceof EntityCollisionContext entitycollisioncontext && entitycollisioncontext.getEntity() instanceof AlienEntity
            ? ALIEN_LAYERS_TO_SHAPE.get(
                state.getValue(LAYERS)
            )
            : LAYERS_TO_SHAPE.get(state.getValue(LAYERS));
    }

    @Override
    public @NotNull VoxelShape getBlockSupportShape(BlockState state, @NotNull BlockGetter world, @NotNull BlockPos pos) {
        return LAYERS_TO_SHAPE.get(state.getValue(LAYERS));
    }

    @Override
    public @NotNull VoxelShape getVisualShape(
        BlockState state,
        @NotNull BlockGetter world,
        @NotNull BlockPos pos,
        @NotNull CollisionContext context
    ) {
        return LAYERS_TO_SHAPE.get(state.getValue(LAYERS));
    }

    @Override
    public boolean useShapeForLightOcclusion(@NotNull BlockState state) {
        return true;
    }

    @Override
    public boolean canSurvive(@NotNull BlockState state, LevelReader world, BlockPos pos) {
        var blockState = world.getBlockState(pos.below());
        var isIce = blockState.is(Blocks.ICE);
        var isPackedIce = blockState.is(Blocks.PACKED_ICE);
        var isBarrier = blockState.is(Blocks.BARRIER);
        var isHoney = blockState.is(Blocks.HONEY_BLOCK);
        var isSoulSand = blockState.is(Blocks.SOUL_SAND);
        if (!isIce && !isPackedIce && !isBarrier) {
            if (!isHoney && !isSoulSand)
                return isFaceFull(blockState.getCollisionShape(world, pos.below()), Direction.UP) || blockState.is(
                    this
                ) && blockState.getValue(LAYERS) == 8;
            else
                return true;
        } else
            return false;
    }

    @Override
    public @NotNull BlockState updateShape(
        BlockState state,
        @NotNull Direction direction,
        @NotNull BlockState neighborState,
        @NotNull LevelAccessor world,
        @NotNull BlockPos pos,
        @NotNull BlockPos neighborPos
    ) {
        return !state.canSurvive(world, pos)
            ? Blocks.AIR.defaultBlockState()
            : super.updateShape(
                state,
                direction,
                neighborState,
                world,
                pos,
                neighborPos
            );
    }

    @Override
    public boolean canBeReplaced(BlockState state, BlockPlaceContext context) {
        int layers = state.getValue(LAYERS);
        if (context.getItemInHand().is(asItem()) && layers < 8)
            if (context.replacingClickedOnBlock())
                return context.getClickedFace() == Direction.UP;
            else
                return true;
        else
            return layers == 1;
    }

    @Nullable
    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        var blockState = ctx.getLevel().getBlockState(ctx.getClickedPos());
        if (blockState.is(this))
            return blockState.setValue(LAYERS, Math.min(8, blockState.getValue(LAYERS) + 1));
        else
            return super.getStateForPlacement(ctx);
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(LAYERS);
    }
}

File: src/main/kotlin/com/bvanseg/gigeresque/common/block/NestResinBlock.kt
//package com.bvanseg.gigeresque.common.block
//
//import com.bvanseg.gigeresque.common.entity.AlienEntity
//import com.bvanseg.gigeresque.common.util.clamp
//import net.minecraft.block.*
//import net.minecraft.block.Blocks
//import net.minecraft.entity.Entity
//import net.minecraft.entity.ai.pathing.NavigationType
//import net.minecraft.item.ItemPlacementContext
//import net.minecraft.state.StateManager
//import net.minecraft.state.property.IntProperty
//import net.minecraft.state.property.Properties
//import net.minecraft.util.math.BlockPos
//import net.minecraft.util.math.Direction
//import net.minecraft.util.math.Vec3d
//import net.minecraft.util.shape.VoxelShape
//import net.minecraft.util.shape.VoxelShapes
//import net.minecraft.world.BlockView
//import net.minecraft.world.World
//import net.minecraft.world.WorldAccess
//import net.minecraft.world.WorldView
//
///**
// * @author Boston Vanseghi
// */
//class NestResinBlock(settings: Settings) : Block(settings) {
//
//    companion object {
//        private fun interpolateShapes(divide: Boolean): List<VoxelShape> {
//            val list = mutableListOf<VoxelShape>()
//            for (i in 1..8) {
//                val minY = if (divide) (i * 2.0) / 2.0 else i * 2.0
//                list.add(createCuboidShape(0.0, 0.0, 0.0, 16.0, minY, 16.0))
//            }
//            return list
//        }
//
//        val LAYERS: IntProperty = Properties.LAYERS
//
//        val ALIEN_LAYERS_TO_SHAPE: Array<VoxelShape> = arrayOf(
//            VoxelShapes.empty(),
//            *interpolateShapes(false).toTypedArray()
//        )
//
//        val LAYERS_TO_SHAPE: Array<VoxelShape> = arrayOf(
//            VoxelShapes.empty(),
//            *interpolateShapes(true).toTypedArray()
//        )
//    }
//
//    init {
//        defaultState = (stateManager.defaultState as BlockState).with(LAYERS, 1) as BlockState
//    }
//
//    override fun canPathfindThrough(
//        state: BlockState,
//        world: BlockView,
//        pos: BlockPos,
//        type: NavigationType
//    ): Boolean = when (type) {
//        NavigationType.LAND -> true
//        else -> false
//    }
//
//    override fun getOutlineShape(
//        state: BlockState,
//        world: BlockView,
//        pos: BlockPos,
//        context: ShapeContext
//    ): VoxelShape = ALIEN_LAYERS_TO_SHAPE[(state.get(LAYERS) as Int)]
//
//    override fun getCollisionShape(
//        state: BlockState,
//        world: BlockView,
//        pos: BlockPos,
//        context: ShapeContext
//    ): VoxelShape = if (context is EntityShapeContext && context.entity is AlienEntity) {
//        ALIEN_LAYERS_TO_SHAPE[(state.get(LAYERS) as Int)]
//    } else {
//        LAYERS_TO_SHAPE[(state.get(LAYERS) as Int)]
//    }
//
//    override fun getSidesShape(
//        state: BlockState,
//        world: BlockView,
//        pos: BlockPos
//    ): VoxelShape = LAYERS_TO_SHAPE[(state.get(LAYERS) as Int)]
//
//    override fun getCameraCollisionShape(
//        state: BlockState,
//        world: BlockView,
//        pos: BlockPos,
//        context: ShapeContext
//    ): VoxelShape = LAYERS_TO_SHAPE[(state.get(LAYERS) as Int)]
//
//    override fun hasSidedTransparency(state: BlockState): Boolean = true
//
//    override fun canPlaceAt(state: BlockState, world: WorldView, pos: BlockPos): Boolean {
//        val blockState = world.getBlockState(pos.down())
//        return if (!blockState.isOf(Blocks.ICE) && !blockState.isOf(Blocks.PACKED_ICE) && !blockState.isOf(Blocks.BARRIER)) {
//            if (!blockState.isOf(Blocks.HONEY_BLOCK) && !blockState.isOf(Blocks.SOUL_SAND)) {
//                isFaceFullSquare(
//                    blockState.getCollisionShape(world, pos.down()),
//                    Direction.UP
//                ) || blockState.isOf(this) && blockState.get(
//                    LAYERS
//                ) as Int == 8
//            } else {
//                true
//            }
//        } else {
//            false
//        }
//    }
//
//    override fun getStateForNeighborUpdate(
//        state: BlockState,
//        direction: Direction,
//        neighborState: BlockState,
//        world: WorldAccess,
//        pos: BlockPos,
//        neighborPos: BlockPos
//    ): BlockState = if (!state.canPlaceAt(world, pos)) {
//        Blocks.AIR.defaultState
//    } else {
//        super.getStateForNeighborUpdate(
//            state,
//            direction,
//            neighborState,
//            world,
//            pos,
//            neighborPos
//        )
//    }
//
//    override fun canReplace(state: BlockState, context: ItemPlacementContext): Boolean {
//        val i = state.get(LAYERS) as Int
//        return if (context.stack.isOf(asItem()) && i < 8) {
//            if (context.canReplaceExisting()) {
//                context.side == Direction.UP
//            } else {
//                true
//            }
//        } else {
//            i == 1
//        }
//    }
//
//    override fun getPlacementState(ctx: ItemPlacementContext): BlockState? {
//        val blockState = ctx.world.getBlockState(ctx.blockPos)
//        return if (blockState.isOf(this)) {
//            val i = blockState.get(LAYERS) as Int
//            blockState.with(LAYERS, 8.coerceAtMost(i + 1)) as BlockState
//        } else {
//            super.getPlacementState(ctx)
//        }
//    }
//
//    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
//        builder.add(LAYERS)
//    }
//
//    override fun onEntityCollision(state: BlockState, world: World, pos: BlockPos, entity: Entity) {
//        if (entity !is AlienEntity) {
//            val multiplier = clamp(1.0 / (state.get(LAYERS)), 0.0, 1.0)
//            entity.slowMovement(state, Vec3d(1.0 * multiplier, 1.0, 1.0 * multiplier))
//        }
//    }
//}